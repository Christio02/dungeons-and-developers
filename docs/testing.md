# Testing

## How to run tests

### E2E testing

For E2E testing we have used Cypress to simulate different user stories. We have made;

- **Login.cy**: Tests login, searching, and user interaction with data
- **Filtering.cy**: tests login and filtering.

To run the E2E tests you need a running frontend and backend, be in the frontend folder and run:

```bash
  npx cypress open
```

This should open a Cypress desktop where you need to navigate to E2E testing.
From here you can select the tests and watch as they are ran in a virtual desktop!

### Component testing

We have also made several component testing in the _tests_ folder in our frontend folder. These are to make sure the component
renders correctly, and behaves as expected during interactions.

We have used _AuthContext_ and _AuthProvider_.

We have made sure to use _MockedProvider_ to mock requests, and the components are tested for correct data handling from GraphQL-requests

#### Running the tests

Be in the frontend folder, and you can run:

```bash
  npm test
```

This document outlines the testing strategy employed for our project, detailing the rationale behind our choices, the tools used, and the achieved test coverage.

## Why We Test

Thorough testing is crucial to ensure the reliability, maintainability, and correctness of our application. It helps us catch bugs early in the development cycle, catch edge-cases, and provides confidence in the codebase's stability.

### Testing Approach

Our testing strategy employs a multi-layered approach, combining both component and end-to-end (E2E) testing:

- **Component Testing (Vitest):** We utilize Vitest for unit and component testing, focusing on individual components in isolation. This allows us to verify the functionality of smaller units of code, ensuring they behave as expected independently. We prioritize testing UI components with Vitest, that are not too complex. Early we discovered that large components, and interactive components needed too much set-up with mock-data and contexts. Therfore we have excluded the large components/pages from this type of testing. We also leveraged a coverage reporter, to identify un-tested components.
  For `Counter` component we had to use fireEvent for user interaction, because the component requires user to hold down mouse/key on keyboard, to increment/decrement score.

- **End-to-End Testing (Cypress):** For larger components/pages and critical user flows, we leverage Cypress for E2E testing. This approach tests the application as a whole, simulating real user interactions within a browser environment. E2E tests help us identify integration issues and ensure the application functions correctly from the user's perspective. We defined some user stories to follow, to make sure the e2e testing followed a logical flow from a user's perspective.

### Testing Tools

- **Vitest:** A fast and efficient unit testing framework that integrates seamlessly with Vite projects.
- **User Events:** A library that provides utilities for simulating user interactions (e.g., clicks, typing) within tests.
- **Snapshot Testing:** A technique that captures the rendered output of a component and compares it against a previously saved "snapshot." This helps detect unexpected UI changes.
- **Cypress:** A comprehensive E2E testing framework that provides a robust and user-friendly environment for testing web applications.

### Coverage

We strive for a test coverage that covers the most critical parts of our website. The flow and branches a user can take when entering our website. As stated earlier we have excluded some components/pages from report and vitest, because we considered it too difficult and time consuming.

**[Insert Coverage Report Here]**

**Tested Files:**

- **Component Tests:** [List of tested component files]
- **E2E Tests:** [List of tested user flows/scenarios]

### Rationale for Testing Choices

We opted for this combined approach to leverage the strengths of both component and E2E testing:

- **Component tests** allow for rapid feedback and efficient testing of individual units of code. They are particularly useful for testing complex logic and interactions within smaller components.
- **E2E tests** provide a higher-level view of the application, ensuring all components work together correctly and the user experience is seamless.

By combining these methods, we achieve a comprehensive testing strategy that balances speed, efficiency, and confidence in our application's quality.
