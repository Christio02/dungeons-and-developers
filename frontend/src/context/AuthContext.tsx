import { createContext, ReactNode, useReducer } from 'react';
import { LoginState, LoginAction, LogoutAction } from '../interfaces/AuthContextProps.ts';

/**
 * AuthContext
 *
 * Provides authentication context to manage user login, logout, and related state.
 * - Stores user information such as `token`, `userId`, and `userName`.
 * - Persists authentication data in `localStorage`.
 * - Allows components to access and update authentication state via context.
 *
 * Features:
 * - Implements a `useReducer` for managing authentication state transitions.
 * - Includes `LOGIN` and `LOGOUT` actions to update state and storage.
 * - Provides helper functions for login and logout logic.
 *
 * Components:
 * - `AuthProvider`: Wraps application components to provide authentication context.
 * - `authReducer`: Reducer function for managing state transitions.
 * - `AuthContext`: Context object for accessing authentication data and methods.
 *
 * Usage:
 * - Wrap the application root with `<AuthProvider>` to enable authentication management.
 * - Access context using `useContext(AuthContext)` in child components.
 */

const initialState: LoginState = {
  token: localStorage.getItem('token') || '',
  userId: localStorage.getItem('userId') || '',
  userName: localStorage.getItem('userName') || '',
};

type AuthAction = LoginAction | LogoutAction;

const AuthContext = createContext<{
  token: string;
  userId: string;
  userName: string;
  login: (data: { token: string; userId: string; userName: string }) => void;
  logout: () => void;
}>({
  token: '',
  userId: '',
  userName: '',
  login: () => {},
  logout: () => {},
});

function authReducer(state: LoginState, action: AuthAction): LoginState {
  switch (action.type) {
    case 'LOGIN':
      return {
        token: action.payload.token,
        userId: action.payload.userId,
        userName: action.payload.userName,
      };
    case 'LOGOUT':
      return {
        token: '',
        userId: '',
        userName: '',
      };
    default:
      return state;
  }
}

function AuthProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(authReducer, initialState);

  const login = (userData: { token: string; userId: string; userName: string }) => {
    console.log('Logging in with:', userData);
    localStorage.setItem('token', userData.token);
    localStorage.setItem('userId', userData.userId);
    localStorage.setItem('userName', userData.userName);
    localStorage.setItem('loginToast', 'true');
    dispatch({ type: 'LOGIN', payload: userData });
  };

  const logout = () => {
    localStorage.clear();
    dispatch({ type: 'LOGOUT' });
  };

  return (
    <AuthContext.Provider value={{ token: state.token, userId: state.userId, userName: state.userName, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext, AuthProvider };
