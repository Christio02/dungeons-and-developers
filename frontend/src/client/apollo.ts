import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import { ReviewType } from '../interfaces/ReviewProps.ts';

const httpLink = createHttpLink({
  uri: 'http://localhost:4000/graphql',
});

const authLink = setContext((_, { headers }) => {
  const token = localStorage.getItem('token');
  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  devtools: {
    enabled: true,
  },
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          user: {
            keyArgs: ['id'],
            merge(existing = {}, incoming) {
              return { ...existing, ...incoming };
            },
          },
        },
      },
      User: {
        keyFields: ['id'],
        fields: {
          favoritedMonsters: {
            merge: false,
          },
          dungeonName: {
            merge: false,
          },
          class: {
            merge(_, incoming) {
              return incoming;
            },
          },

          race: {
            merge(_, incoming) {
              return incoming;
            },
          },
          equipments: {
            merge(_, incoming) {
              return incoming;
            },
          },
        },
      },
      Class: {
        keyFields: ['name'],
      },
      Race: {
        keyFields: ['name'],
      },
      Ability: {
        keyFields: ['id'],
      },
      Equipment: {
        keyFields: ['name'],
      },
      Monster: {
        keyFields: ['id'],
        fields: {
          reviews: {
            merge(existing = [], incoming) {
              const mergedReviews = [...existing];
              incoming.forEach((incomingReview: ReviewType) => {
                if (!mergedReviews.some((review) => review.id === incomingReview.id)) {
                  mergedReviews.push(incomingReview);
                }
              });
              return mergedReviews;
            },
          },
        },
      },
      Review: {
        keyFields: ['id'],
      },
    },
  }),
});

export default client;
