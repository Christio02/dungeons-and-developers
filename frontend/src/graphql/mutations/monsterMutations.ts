import { gql } from '@apollo/client';

export const ADD_REVIEW = gql`
  mutation AddReview($monsterId: ID!, $review: ReviewInput!) {
    addReview(monsterId: $monsterId, review: $review) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
export const DELETE_REVIEW = gql`
  mutation DeleteReview($monsterId: ID!, $reviewId: ID!) {
    deleteReview(monsterId: $monsterId, reviewId: $reviewId) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
export const UPDATE_REVIEW = gql`
  mutation UpdateReview($monsterId: ID!, $reviewId: ID!, $review: ReviewInput!) {
    updateReview(monsterId: $monsterId, reviewId: $reviewId, review: $review) {
      id
      user {
        id
        userName
      }
      difficulty
      description
      createdAt
    }
  }
`;
