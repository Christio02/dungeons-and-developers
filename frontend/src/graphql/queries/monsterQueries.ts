import { gql } from '@apollo/client';

/**
 * Monster Queries
 *
 * Features:
 * - Fetches paginated and filtered lists of monsters.
 * - Retrieves monster type counts and hit point ranges for filters.
 * - Provides monster suggestions for search and autocomplete.
 * - Fetches reviews associated with a specific monster.
 */

/**
 * - `GET_MONSTERS`: Fetches a paginated list of monsters based on various filter options.
 *   - Query Parameters:
 *     - `searchTerm` (optional): Filter monsters by name or related terms.
 *     - `offset` (optional): Number of records to skip for pagination.
 *     - `limit` (optional): Maximum number of records to retrieve.
 *     - `types` (optional): Filter by a list of monster types.
 *     - `minHp` (optional): Minimum hit points for filtering.
 *     - `maxHp` (optional): Maximum hit points for filtering.
 *     - `sortOption` (optional): Sorting option for monsters.
 *   - Query Response:
 *     - `monsters`: List of monster objects with:
 *       - `id`: Unique identifier.
 *       - `name`: Name of the monster.
 *       - `size`: Monster size category.
 *       - `type`: Type of monster.
 *       - `alignment`: Alignment of the monster.
 *       - `hit_points`: Total hit points.
 *       - `image`: URL of the monster's image.
 *       - `averageDifficulty`: Average difficulty rating.
 *       - `reviewsCount`: Number of reviews.
 *     - `totalMonsters`: Total number of monsters available.
 *     - `minHp` and `maxHp`: Minimum and maximum hit points across all monsters.
 */

export const GET_MONSTERS = gql`
  query GetMonsters(
    $searchTerm: String
    $offset: Int
    $limit: Int
    $types: [String!]
    $minHp: Int
    $maxHp: Int
    $sortOption: String
  ) {
    monsters(
      searchTerm: $searchTerm
      offset: $offset
      limit: $limit
      types: $types
      minHp: $minHp
      maxHp: $maxHp
      sortOption: $sortOption
    ) {
      monsters {
        id
        name
        size
        type
        alignment
        hit_points
        image
        averageDifficulty
        reviewsCount
      }
      totalMonsters
      minHp
      maxHp
    }
  }
`;

/**
 * - `GET_MONSTER_TYPE_COUNTS`: Fetches counts of monster types for given hit point ranges.
 *   - Query Parameters:
 *     - `minHp` (optional): Minimum hit points for filtering.
 *     - `maxHp` (optional): Maximum hit points for filtering.
 *   - Query Response:
 *     - `monsterTypeCounts`: Array of types with:
 *       - `type`: Monster type.
 *       - `count`: Count of monsters of this type.
 */
export const GET_MONSTER_TYPE_COUNTS = gql`
  query GetMonsterTypeCounts($minHp: Int, $maxHp: Int) {
    monsterTypeCounts(minHp: $minHp, maxHp: $maxHp) {
      type
      count
    }
  }
`;

/**
 * - `GET_MONSTER_HP_RANGE`: Retrieves the global range of hit points for all monsters.
 *   - Query Response:
 *     - `monsterHpRange`: Object with:
 *       - `minHp`: Minimum hit points.
 *       - `maxHp`: Maximum hit points.
 */
export const GET_MONSTER_HP_RANGE = gql`
  query GetMonsterHpRange {
    monsterHpRange {
      minHp
      maxHp
    }
  }
`;

/**
 * - `GET_MONSTER_SUGGESTIONS`: Provides monster suggestions for autocomplete or filtering.
 *   - Query Parameters:
 *     - `searchTerm` (optional): Filter suggestions by term.
 *     - `types` (optional): Restrict suggestions to specific monster types.
 *     - `minHp` (optional): Minimum hit points for filtering.
 *     - `maxHp` (optional): Maximum hit points for filtering.
 *   - Query Response:
 *     - `monsters`: List of suggested monsters with:
 *       - `id`: Unique identifier.
 *       - `name`: Monster name.
 */
export const GET_MONSTER_SUGGESTIONS = gql`
  query GetMonsterSuggestions($searchTerm: String, $types: [String!], $minHp: Int, $maxHp: Int) {
    monsters(searchTerm: $searchTerm, types: $types, minHp: $minHp, maxHp: $maxHp) {
      monsters {
        id
        name
      }
    }
  }
`;

/**
 * - `GET_MONSTER_REVIEWS`: Fetches reviews associated with a specific monster.
 *   - Query Parameters:
 *     - `monsterId` (required): Unique identifier of the monster.
 *   - Query Response:
 *     - `monster`: Object containing:
 *       - `id`: Monster identifier.
 *       - `reviews`: List of review objects with:
 *         - `id`: Review identifier.
 *         - `user`: Reviewer details with:
 *           - `id`: User identifier.
 *           - `userName`: Name of the reviewer.
 *         - `difficulty`: Difficulty rating of the monster.
 *         - `description`: Text of the review.
 *         - `createdAt`: Timestamp of the review creation.
 */
export const GET_MONSTER_REVIEWS = gql`
  query GetMonsterReviews($monsterId: ID!) {
    monster(id: $monsterId) {
      id
      reviews {
        id
        user {
          id
          userName
        }
        difficulty
        description
        createdAt
      }
    }
  }
`;
