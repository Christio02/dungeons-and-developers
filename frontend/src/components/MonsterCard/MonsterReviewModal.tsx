import { useMutation, useQuery } from '@apollo/client';
import { Box, Button, Dialog, DialogActions, DialogContent, DialogContentText } from '@mui/material';
import React, { useContext, useEffect, useState } from 'react';
import { AuthContext } from '../../context/AuthContext';
import { ADD_REVIEW, UPDATE_REVIEW } from '../../graphql/mutations/monsterMutations.ts';
import { GET_MONSTER_REVIEWS } from '../../graphql/queries/monsterQueries.ts';
import { useToast } from '../../hooks/useToast';
import { AddReviewData, GetMonsterReviewsData, UpdateReviewData } from '../../interfaces/MonsterCardProps.ts';
import { ReviewFormType, ReviewType } from '../../interfaces/ReviewProps.ts';
import ReviewSlider from './ReviewSlider.tsx';
import ReviewTextField from './ReviewTextField.tsx';

const MonsterReviewModal = ({ name, monsterId, image }: ReviewFormType) => {
  const { userId } = useContext(AuthContext);
  const [isOpen, setIsOpen] = useState(false);
  const [editMode, setEditMode] = useState(false);
  const [difficulty, setDifficulty] = useState(50);
  const [description, setDescription] = useState('');
  const [originalDifficulty, setOriginalDifficulty] = useState(50);
  const [originalDescription, setOriginalDescription] = useState('');
  const { showToast } = useToast();

  const { data } = useQuery<GetMonsterReviewsData>(GET_MONSTER_REVIEWS, {
    variables: { monsterId },
    skip: !isOpen,
    fetchPolicy: 'network-only',
  });
  const existingReview = data?.monster?.reviews.find((review: ReviewType) => review.user.id === userId);

  const [addReview] = useMutation<AddReviewData>(ADD_REVIEW, {
    update(cache, { data }) {
      if (!data) return;
      const existingData = cache.readQuery<GetMonsterReviewsData>({
        query: GET_MONSTER_REVIEWS,
        variables: { monsterId },
      });
      if (existingData && existingData.monster) {
        const newReviews = data.addReview.reviews;

        cache.writeQuery<GetMonsterReviewsData>({
          query: GET_MONSTER_REVIEWS,
          variables: { monsterId },
          data: {
            monster: {
              ...existingData.monster,
              reviews: newReviews,
            },
          },
        });
      }
    },
  });

  const [updateReview] = useMutation<UpdateReviewData>(UPDATE_REVIEW, {
    update(cache, { data }) {
      if (!data) return;

      const existingData = cache.readQuery<GetMonsterReviewsData>({
        query: GET_MONSTER_REVIEWS,
        variables: { monsterId },
      });

      if (existingData && existingData.monster) {
        const updatedReview = data.updateReview;

        const updatedReviews = existingData.monster.reviews.filter((review) => review.id !== updatedReview.id);
        updatedReviews.push(updatedReview);

        cache.writeQuery<GetMonsterReviewsData>({
          query: GET_MONSTER_REVIEWS,
          variables: { monsterId },
          data: {
            monster: {
              ...existingData.monster,
              reviews: updatedReviews,
            },
          },
        });
      }
    },
  });

  useEffect(() => {
    if (existingReview) {
      setDescription(existingReview.description);
      setDifficulty(existingReview.difficulty);
      setOriginalDescription(existingReview.description);
      setOriginalDifficulty(existingReview.difficulty);
      setEditMode(true);
    } else {
      // reset form if no review
      setDescription('');
      setDifficulty(50);
      setOriginalDescription('');
      setOriginalDifficulty(50);
    }
  }, [existingReview]);

  const handleClickOpen = () => {
    setIsOpen(true);
    if (existingReview) {
      setEditMode(true);
    } else {
      setDescription('');
      setDifficulty(50);
      setEditMode(false);
    }
  };

  const handleClose = () => {
    if (editMode) {
      setDescription(originalDescription);
      setDifficulty(originalDifficulty);
    } else {
      setDescription('');
      setDifficulty(50);
    }
    setIsOpen(false);
    setEditMode(false);
  };

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();
    if (!userId) {
      showToast({ message: 'You must be logged in to submit a review', type: 'error', duration: 3000 });
      return;
    }

    try {
      if (editMode && existingReview && existingReview.id) {
        // Update existing review
        await updateReview({
          variables: {
            monsterId,
            reviewId: existingReview.id,
            review: { user: userId, difficulty, description },
          },
        });
        showToast({ message: `Review on ${name} updated`, type: 'success', duration: 3000 });
      } else {
        await addReview({
          variables: {
            monsterId,
            review: { user: userId, difficulty, description },
          },
        });
        showToast({ message: `Review on ${name} submitted`, type: 'success', duration: 3000 });
      }
      setIsOpen(false);
    } catch (error) {
      console.error('Error submitting review:', error);
      showToast({ message: `Failed to submit review on ${name}`, type: 'error', duration: 3000 });
    }
  };

  return (
    <>
      <Button
        variant="outlined"
        className="review-button"
        onClick={handleClickOpen}
        sx={{
          color: 'white',
          borderColor: '#DB3232',
          backgroundColor: '#DB3232',
          padding: '0.2vw 0.5vw',
          minWidth: 'unset',
          minHeight: 'unset',
          height: 'auto',
          lineHeight: 'normal',
          borderRadius: '4px',
          boxSizing: 'border-box',
          '&:hover': {
            borderColor: '#DB3232',
            backgroundColor: 'black',
            color: '#DB3232',
          },
          fontFamily: 'MedievalSharp',
          fontSize: {
            xs: '3.5vw',
            sm: '2vw',
            md: '1.3vw',
            lg: '1.2vw',
            xl: '1vw',
          },
          textTransform: 'none',
        }}
        aria-label="Review button"
      >
        Review
      </Button>
      <Dialog
        open={isOpen}
        onClose={handleClose}
        PaperProps={{
          component: 'form',
          sx: {
            width: { xs: '100vw', sm: '95vw', md: '90vw', lg: '80vw', xl: '70vw' },
            height: { xs: '100vh', sm: '85vh', md: '85vh', lg: '60vh', xl: '70vh' },
            maxWidth: 'none',
            padding: { xs: 0, sm: 4 },
            backgroundColor: 'black',
          },
        }}
      >
        <DialogContent className="flex flex-col xl:flex-row xl:items-center bg-black gap-6">
          <Box className="w-full xl:w-1/2 flex flex-col gap-[50px] xl:items-center">
            <img src={image} alt="Image of selected monster" className="w-full sm:w-3/4 xl:w-1/2 rounded" />
            <h2 className="sub-header font-bold">Review of {name}</h2>
          </Box>
          <article className="flex flex-col gap-4 w-full xl:w-1/2">
            <DialogContentText sx={{ color: 'white', fontSize: '1.5rem', fontFamily: 'MedievalSharp' }}>
              Difficulty
            </DialogContentText>
            <ReviewSlider value={difficulty} onChange={(_, value) => setDifficulty(value as number)} />
            <ReviewTextField value={description} onChange={(e) => setDescription(e.target.value.slice(0, 300))} />
          </article>
        </DialogContent>

        <DialogActions className="bg-black">
          <Button
            onClick={handleClose}
            aria-label="Cancel-button"
            sx={{
              color: 'white',
              borderColor: 'white',
              '&:hover': {
                borderColor: '#DB3232',
                color: '#DB3232',
              },
              fontFamily: 'MedievalSharp',
              fontSize: '1.5rem',
            }}
          >
            Cancel
          </Button>
          <Button
            onClick={handleSubmit}
            type="submit"
            aria-label="Save-button"
            sx={{
              color: 'white',
              borderColor: 'white',
              '&:hover': {
                borderColor: '#DB3232',
                color: '#DB3232',
              },
              fontFamily: 'MedievalSharp',
              fontSize: '1.5rem',
            }}
          >
            {editMode ? 'UPDATE' : 'SUBMIT'}
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default MonsterReviewModal;
