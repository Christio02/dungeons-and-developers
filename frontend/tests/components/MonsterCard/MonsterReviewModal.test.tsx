import { MockedProvider } from '@apollo/client/testing';
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import MonsterReviewModal from '../../../src/components/MonsterCard/MonsterReviewModal.tsx';
import { AuthContext } from '../../../src/context/AuthContext.tsx';
import { GET_MONSTER_REVIEWS } from '../../../src/graphql/queries/monsterQueries.ts';
import { ADD_REVIEW, UPDATE_REVIEW } from '../../../src/graphql/mutations/monsterMutations.ts';

export type Review = {
  id: string;
  user: {
    id: string;
    userName: string;
  };
  difficulty: number;
  description: string;
  createdAt: string;
};
// Mock the useToast hook
vi.mock('../../../src/hooks/useToast.ts', () => ({
  useToast: () => ({
    showToast: vi.fn(),
  }),
}));

const mockProps = {
  name: 'Dragon',
  monsterId: '123',
  image: 'dragon.jpg',
};

const mockUserId = 'user123';
const mockToken = 'mock-token';
const mockUsername = 'bob';
const mockLogin = vi.fn();
const mockLogout = vi.fn();

const initialEmptyReviews: Review[] = [];
const mockReviewsUpdate: Review[] = [
  {
    id: 'review1',
    user: { id: mockUserId, userName: mockUsername },
    difficulty: 75,
    description: 'Updated review',
    createdAt: '2024-11-05T12:00:00Z',
  },
];

const mockReviewsAdd: Review[] = [
  {
    id: 'review1',
    user: { id: mockUserId, userName: mockUsername },
    difficulty: 50,
    description: 'New review',
    createdAt: '2024-11-05T12:00:00Z',
  },
];

// Apollo query mocks
const mocksForAddingReview = [
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          id: mockProps.monsterId,
          __typename: 'Monster',
          reviews: initialEmptyReviews,
        },
      },
    },
  },
  {
    request: {
      query: ADD_REVIEW,
      variables: {
        monsterId: mockProps.monsterId,
        review: {
          user: mockUserId,
          difficulty: 50,
          description: 'New review',
        },
      },
    },
    result: {
      data: {
        addReview: {
          id: mockProps.monsterId,
          __typename: 'Monster',
          reviews: mockReviewsAdd.map((review) => ({
            ...review,
            __typename: 'Review',
            user: { ...review.user, __typename: 'User' },
          })),
        },
      },
    },
  },
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          id: mockProps.monsterId,
          __typename: 'Monster',
          reviews: mockReviewsAdd.map((review) => ({
            ...review,
            __typename: 'Review',
            user: { ...review.user, __typename: 'User' },
          })),
        },
      },
    },
  },
];

const mocksForUpdatingReview = [
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          id: mockProps.monsterId,
          __typename: 'Monster',
          reviews: mockReviewsAdd,
        },
      },
    },
  },
  {
    request: {
      query: UPDATE_REVIEW,
      variables: {
        monsterId: mockProps.monsterId,
        reviewId: 'review1',
        review: {
          user: mockUserId,
          difficulty: 50,
          description: 'Updated review',
        },
      },
    },
    result: {
      data: {
        updateReview: {
          id: 'review1',
          user: {
            id: mockUserId,
            userName: mockUsername,
          },
          difficulty: 75,
          description: 'Updated review',
          createdAt: '2024-11-06T08:00:00Z',
        },
      },
    },
  },
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          id: mockProps.monsterId,
          reviews: mockReviewsUpdate,
        },
      },
    },
  },
];
const mocksForResetForm = [
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          id: mockProps.monsterId,
          __typename: 'Monster',
          reviews: mockReviewsAdd,
        },
      },
    },
  },
  {
    request: {
      query: GET_MONSTER_REVIEWS,
      variables: { monsterId: mockProps.monsterId },
    },
    result: {
      data: {
        monster: {
          id: mockProps.monsterId,
          __typename: 'Monster',
          reviews: mockReviewsAdd,
        },
      },
    },
  },
];

const renderComponent = (mocks: any) => {
  // any since structure is not always the same
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <AuthContext.Provider
        value={{ userId: mockUserId, userName: mockUsername, logout: mockLogout, login: mockLogin, token: mockToken }}
      >
        <MonsterReviewModal {...mockProps} />
      </AuthContext.Provider>
    </MockedProvider>
  );
};

describe('MonsterReviewModal', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders review button correctly', async () => {
    const { container } = renderComponent(mocksForAddingReview);
    const reviewButton = await screen.findByText('Review');
    expect(reviewButton).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('opens modal when review button is clicked', async () => {
    renderComponent(mocksForAddingReview);
    const user = userEvent.setup();

    const reviewButton = await screen.findByText('Review');
    expect(reviewButton).toBeInTheDocument();

    await user.click(reviewButton);

    expect(screen.getByText(`Review of ${mockProps.name}`)).toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute('src', mockProps.image);
    expect(screen.getByText('Difficulty')).toBeInTheDocument();
  });

  it('shows 0 reviews when there are no reviews', async () => {
    renderComponent(mocksForAddingReview);
    const user = userEvent.setup();

    const reviewButton = await screen.findByText('Review');
    expect(reviewButton).toBeInTheDocument();

    await user.click(reviewButton);
    await waitFor(() => {
      expect(screen.getByRole('textbox')).toHaveValue('');
    });
  });

  it('handles submission of new review, and shows it', async () => {
    renderComponent(mocksForAddingReview);

    const user = userEvent.setup();

    const reviewButton = await screen.findByText('Review');
    expect(reviewButton).toBeInTheDocument();

    await user.click(reviewButton);

    await screen.findByRole('dialog');

    const textarea = await screen.findByLabelText('Description');

    await user.clear(textarea);
    await user.type(textarea, 'New review');

    const saveButton = await screen.findByRole('button', { name: /save-button/i });
    await user.click(saveButton);

    await user.click(reviewButton);

    expect(screen.getByLabelText('Description')).toHaveValue('New review');
  });

  it('handles updating existing review', async () => {
    renderComponent(mocksForUpdatingReview);
    const user = userEvent.setup();

    const reviewButton = await screen.findByText('Review');
    expect(reviewButton).toBeInTheDocument();

    await user.click(reviewButton);
    await waitFor(() => {
      expect(screen.getByLabelText('Description')).toHaveValue('New review');
    });

    await user.clear(screen.getByLabelText('Description'));

    await user.type(screen.getByLabelText('Description'), 'Updated review');

    await user.click(screen.getByRole('button', { name: 'Save-button' }));

    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    await user.click(reviewButton);

    await waitFor(() => {
      expect(screen.getByLabelText('Description')).toHaveValue('Updated review');
    });
  });

  it('closes modal when cancel is clicked', async () => {
    renderComponent(mocksForAddingReview);
    const user = userEvent.setup();

    const reviewButton = await screen.findByText('Review');
    expect(reviewButton).toBeInTheDocument();

    await user.click(reviewButton);

    await user.click(reviewButton);

    const cancelButton = await screen.findByText('Cancel');
    await user.click(cancelButton);

    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  it(
    'respects character limit for review description',
    async () => {
      renderComponent(mocksForAddingReview);
      const user = userEvent.setup();

      const reviewButton = await screen.findByText('Review');
      expect(reviewButton).toBeInTheDocument();

      await user.click(reviewButton);

      const longText = 'a'.repeat(301);
      await user.clear(screen.getByRole('textbox'));
      await user.type(screen.getByRole('textbox'), longText);

      expect(screen.getByRole('textbox')).toHaveValue('a'.repeat(300));
    },
    { timeout: 10000 }
  );

  it('resets form when modal is closed and re-opened', async () => {
    renderComponent(mocksForResetForm);
    const user = userEvent.setup();

    const reviewButton = await screen.findByText('Review');

    expect(reviewButton).toBeInTheDocument();

    await user.click(reviewButton);

    await waitFor(() => {
      expect(screen.getByLabelText('Description')).toHaveValue('New review');
    });
    await user.clear(screen.getByLabelText('Description'));

    await user.type(screen.getByLabelText('Description'), 'Temporary review');

    await user.click(screen.getByText('Cancel'));

    await user.click(reviewButton);

    await waitFor(() => {
      expect(screen.getByRole('textbox')).toHaveValue('New review');
    });
  });
});
