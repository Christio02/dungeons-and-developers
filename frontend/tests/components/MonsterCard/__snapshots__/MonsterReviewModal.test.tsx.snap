// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MonsterReviewModal > opens modal when review button is clicked 1`] = `
<div>
  <button
    aria-label="Review button"
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1l3bbxn-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Review
  </button>
</div>
`;

exports[`MonsterReviewModal > renders review button correctly 1`] = `
<div>
  <button
    aria-label="Review button"
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary review-button css-1l3bbxn-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Review
  </button>
</div>
`;
